locksmith/src/utils/pricing.ts

```diff
@@ -57,8 +75,176 @@ export async function defiLammaPrice({
+    lockContract.tokenAddress(),
+  ])
+  const decimals =
+    currencyContractAddress &&
+    currencyContractAddress !== ethers.constants.AddressZero
+      ? await getErc20Decimals(currencyContractAddress, provider)
+      : networks[network].nativeCurrency?.decimals || 18
+
+  return {
+    decimals,
+    keyPrice,
+    currencyContractAddress,
+  }
+}
+
+export const getKeyPricingInUSD = async ({
+  recipients,
+  network,
+  lockAddress,
+  data: dataArray,
+  referrers,
+}: KeyPricingOptions) => {
+  const web3Service = new Web3Service(networks)
+  const { keyPrice, decimals, currencyContractAddress } =
+    await getLockKeyPricing({
+      lockAddress,
+      network,
+    })
+
+  const usdPricing = await defiLammaPrice({
+    network,
+    address:
+      !currencyContractAddress ||
+      currencyContractAddress === ethers.constants.AddressZero
+        ? undefined
+        : currencyContractAddress,
+    amount: 1,
+  })
+
+  const defaultPrice = fromDecimal(keyPrice, decimals)
+
+  const defaultPricing = {
+    amount: defaultPrice,
+    decimals,
+    symbol: usdPricing.symbol,
+    amountInUSD: usdPricing?.price
+      ? defaultPrice * usdPricing.price
+      : undefined,
+    amountInCents: usdPricing?.price
+      ? Math.round(defaultPrice * usdPricing.price * 100)
+      : 0,
+  }
+
+  const result = await Promise.all(
+    recipients.map(async (address, index) => {
+      const data = dataArray?.[index] ?? '0x'
+      const referrer = referrers?.[index] ?? address!
+
+      if (!address) {
+        return {
+          recipient: null,
+          ...defaultPricing,
+        }
+      }
+
+      try {
+        const purchasePrice = await web3Service.purchasePriceFor({
+          lockAddress,
+          userAddress: address,
+          data,
+          network,
+          referrer,
+        })
+        const amount = fromDecimal(purchasePrice, decimals)
+        return {
+          address,
+          price: {
+            amount,
+            decimals,
+            symbol: usdPricing.symbol,
+            amountInUSD: usdPricing?.price
+              ? amount * usdPricing.price
+              : undefined,
+            amountInCents: usdPricing?.price
+              ? Math.round(amount * usdPricing.price * 100)
+              : 0,
+          },
+        }
+      } catch (error) {
+        logger.error(error)
+        return {
+          address,
+          ...defaultPricing,
+        }
+      }
+    })
+  )
+  return result
+}
+
+// Fee denominated in cents
+export const getCreditCardProcessingFee = (subtotal: number) => {
+  const serviceFee = getUnlockServiceFee(subtotal)
+  const total = subtotal + serviceFee
+  // This is rounded up to an integer number of cents.
+  const percentageFee = Math.ceil(total * stripePercentage)
+  return baseStripeFee + percentageFee
+}
+
+// Fee denominated in cents
+export const getUnlockServiceFee = (cost: number) => {
+  return Math.ceil(cost * 0.1) // Unlock charges 10% of transaction.
```

Comment:
Maybe add the cost of gas here... ideally as a simulation of the tx to be sent?

Reply:
Added the same gas cost above we have on API.